frontend/src/contexts/campaign.context.tsx

```diff
@@ -19,10 +29,27 @@ const CampaignContextProvider = ({
     new Campaign({}) as SMSCampaign | EmailCampaign | TelegramCampaign
   )
 
+  function updateCampaign(
+    changes: Partial<SMSCampaign | EmailCampaign | TelegramCampaign>
+  ) {
+    setCampaign((c) => {
+      const updatedCampaign = Object.assign(cloneDeep(c), changes) as
+        | SMSCampaign
+        | EmailCampaign
+        | TelegramCampaign
+      updatedCampaign.setProgress()
+      return updatedCampaign
+    })
+  }
+
   return (
     <CampaignContext.Provider
       value={{
         campaign,
+        // updateCampaign is re-created for every re-render of CampaignContextProvider
+        // Memoise updateCampaign to prevent unnecessary renders in subscribers when
+        // defined as dependencies in useEffect
+        updateCampaign: useCallback(updateCampaign, []),
```

Comment:
When `CampaignContextProvider` re-renders due to changes in campaign state, the `updateCampaign` function is re-created. This affects hooks which have `updateCampaign` as a dependency.
```
useEffect(() => {
    updateCampaign({ isCsvProcessing, csvFilename, numRecipients })
}, [isCsvProcessing, csvFilename, numRecipients, updateCampaign])
```
For example, the useEffect hook will execute again as `updatedCampaign`'s value changed. This executes `updateCampaign` within the hook and the context campaign state is updated, `CampaignContextProvider` re-renders .... This would result in infinite re-renders. Hence we are using `useCallback` here to memoize the `updateCampaign` function to prevent unnecessary side effects.

This article helps to elaborate on this concept - https://reacttraining.com/blog/when-to-use-functions-in-hooks-dependency-array/, particularly the section on `useCallback`